### PR TITLE
NDRS-1006: Add timeout-based sync failure.

### PR DESCRIPTION
**Pull request by goral09**
https://casperlabs.atlassian.net/browse/NDRS-1006

Instead of failing when we run out of peers before we manage to synchronize _any_ linear chain block, we will wait up to 5 minutes (same as in consensus). When we run out of peers, after we started syncing blocks, we will fail and exit. The behavior has been modified only for the "initial sync".

As part of the work in this PR, I've also removed some of the `panic!` calls from the `LinearChainSync` and replaced them with calls to `fatal!` macro that should allow the node to drop all components cleanly.

### CODE REVIEW

File: node/src/components/linear_chain_sync.rs [hunk 1]
```diff
@@ -261,11 +278,17 @@ impl<I: Clone + PartialEq + 'static> LinearChainSync<I> {
...
-                        "Block execution result doesn't match received block."
-                    ),
-                    None => panic!("Unexpected block execution results."),
+                    Some(expected) if expected != &block => {
+                        error!(?expected, got=?block, "block execution result doesn't match received block");
+                        return fatal!(effect_builder, "unexpected block execution result")
+                            .ignore();
+                    }
+                    None => {
+                        error!("block execution results recevied when not expected");
```
Comment (afck) [inline node/src/components/linear_chain_sync.rs:287]:
```suggestion
                        error!("block execution results received when not expected");
```

File: node/src/components/linear_chain_sync.rs [hunk 2]
```diff
@@ -280,11 +303,17 @@ impl<I: Clone + PartialEq + 'static> LinearChainSync<I> {
...
```
Comment (afck) [inline node/src/components/linear_chain_sync.rs:312]:
```suggestion
                        error!("block execution results received when not expected");
```

File: node/src/components/linear_chain_sync.rs [hunk 3]
```diff
@@ -569,9 +609,19 @@ where
...
+                                return fatal!(
+                                    effect_builder,
+                                    "failed to synchronize linear chain"
+                                )
+                                .ignore();
+                            }
                             None => {
-                                error!(%block_hash, "Could not download linear block from any of the peers.");
-                                panic!("Failed to download linear chain.")
+                                warn!("run out of peers before managed to start syncing. Resetting peers' list and continueing");
```
Comment (afck) [inline node/src/components/linear_chain_sync.rs:621]:
```suggestion
                                warn!("run out of peers before managed to start syncing. Resetting peers' list and continuing");
```

**Review by afck — APPROVED** [review]

**Comment by goral09** [comment]
```
{"timestamp":"Mar 09 14:36:46.872","level":"DEBUG","fields":{"event":"linear chain: Initialize timeout","q":"Regular"},"target":"casper_node::reactor","span":{"ev":40,"name":"dispatch events"},"spans":[{"ev":40,"name":"crank"},{"ev":40,"name":"dispatch events"}]}
{"timestamp":"Mar 09 14:36:46.872","level":"INFO","fields":{"message":"hasn't downloaded any blocks in expected time window. Shutting down…"},"target":"casper_node::components::linear_chain_sync","span":{"ev":40,"name":"dispatch events"},"spans":[{"ev":40,"name":"crank"},{"ev":40,"name":"dispatch events"}]}
{"timestamp":"Mar 09 14:36:46.872","level":"DEBUG","fields":{"event":"control: fatal error [node/src/components/linear_chain_sync.rs:754]: no syncing progress, shutting down…","q":"Control"},"target":"casper_node::reactor","span":{"ev":41,"name":"dispatch events"},"spans":[{"ev":41,"name":"crank"},{"ev":41,"name":"dispatch events"}]}
{"timestamp":"Mar 09 14:36:46.872","level":"ERROR","fields":{"message":"fatal error via control announcement","file":"node/src/components/linear_chain_sync.rs","line":"754","msg":"no syncing progress, shutting down…"},"target":"casper_node::reactor","span":{"ev":41,"name":"dispatch events"},"spans":[{"ev":41,"name":"crank"},{"ev":41,"name":"dispatch events"}]}
{"timestamp":"Mar 09 14:36:46.876","level":"WARN","fields":{"message":"failed running \"/home/mateusz/work/casperlabs/casper-node/utils/nctl/assets/net-1/nodes/node-1/bin/1_0_0/casper-node\" \"validator\" \"/home/mateusz/work/casperlabs/casper-node/utils/nctl/assets/net-1/nodes/node-1/config/1_0_0/config.toml\"","exit_status":"exit code: 101"},"target":"casper_node_launcher::utils"}
```

Tested, works.

**Pull request merged**